shared-module/_canio/Message.h

```diff
@@ -0,0 +1,47 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020 Jeff Epler for Adafruit Industries
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#pragma once
+
+#include "py/obj.h"
+
+typedef struct {
+    mp_obj_base_t base;
+    int id;
+    size_t size;
+    uint8_t data[8];
+    bool rtr;
+} canio_message_obj_t;
+
+void common_hal_canio_message_construct(canio_message_obj_t *self, int id, void *data, size_t size, bool rtr);
+bool common_hal_canio_message_rtr_get(const canio_message_obj_t *self);
+int common_hal_canio_message_id_get(const canio_message_obj_t *self);
+const void *common_hal_canio_message_data_get(const canio_message_obj_t *self);
+size_t common_hal_canio_message_size_get(const canio_message_obj_t *self);
+void common_hal_canio_message_rtr_set(canio_message_obj_t *self, bool rtr);
+void common_hal_canio_message_id_set(canio_message_obj_t *self, int id);
+void common_hal_canio_message_data_set(canio_message_obj_t *self, const void *data, size_t size);
+void common_hal_canio_message_size_set(canio_message_obj_t *self, size_t size);
```

Comment:
The implementation of "Message" is shared among all ports, so I placed the implementation in shared-module.  Is that correct?  Is the only problem that the related function declarations go in `shared-bindings/<module>/<type>.h` and not in `shared-module`?  I had not figured out this pattern yet.  :frowning:

Reply:
Ya, placing the implementation here is what I'd do. I think you've changed it to how I would do it. 

Anything called by shared-bindings goes in the shared-bindings .h file because it is a shared C api that all implementations must provide. The struct for the type is implementation specific so it either goes in common-hal or shared-modules and is why I prefer getter and setter functions instead of accessing the struct directly.